horovod/ray/utils.py

```diff
@@ -0,0 +1,87 @@
+from typing import List, Optional, Callable, Any, Dict
+from contextlib import contextmanager
+import ray
+
+from horovod.runner.driver import driver_service
+from horovod.ray.driver_service import _driver_fn
+from horovod.runner.util import network
+
+
+class _DiscoveryActor:
+    def execute(self, func: Callable) -> Any:
+        """Executes an arbitrary function on self."""
+        return func(self)
+
+
+@contextmanager
+def _maybe_create_workers(all_host_names: List[str],
+                          existing_workers: Optional[List] = None):
+    should_create = not bool(existing_workers)
+    if should_create:
+        existing_workers = []
+        RemoteDiscoveryActor = ray.remote(_DiscoveryActor)
+        for hostname in all_host_names:
+            node_resource = {f"node:{hostname}": 0.001}
+            temporary_worker = RemoteDiscoveryActor.options(
+                resources=node_resource, num_cpus=0).remote()
+            existing_workers.append(temporary_worker)
+
+    yield existing_workers
```

Comment:
Should this be wrapped in a try-finally in case of exception?

```
try:
    yield existing_workers
finally:
    if should_create:
        for worker in existing_workers:
            del worker
```

Reply:
yeah that sounds good! fixed and rebased.